File: windows/SettingsWindow.py
from PySide6.QtCore import *
from PySide6.QtGui import *
from PySide6.QtWidgets import *

from utils import Settings
from ui import Ui_Settings


class SettingsWindow(QDialog):
  def __init__(self, parent) -> None:
    QDialog.__init__(self, parent)
    self.ui = Ui_Settings()
    self.ui.setupUi(self)

    self.settings = Settings.get_instance()

    self.setWindowTitle("Settings")

    self.setWindowFlag(Qt.FramelessWindowHint, True)
    self.setAttribute(Qt.WA_TranslucentBackground)

    minute_validator = QIntValidator(0, 99)
    second_validator = QRegularExpression(r"^[0-9]|[1-5][0-9]$")

    self.ui.leMinutes.setValidator(minute_validator)
    self.ui.leSeconds.setValidator(
        QRegularExpressionValidator(second_validator))

    self.ui.btnSave.clicked.connect(self.save_settings)
    self.ui.btnCancel.clicked.connect(self.close)
    self.ui.cbHideTitleBar.setChecked(self.settings.auto_hide_title_bar)
    self.ui.leMinutes.setText(str(self.settings.time // 60))
    self.ui.leSeconds.setText(str(self.settings.time % 60))

  def save_settings(self) -> None:
    self.settings.auto_hide_title_bar = self.ui.cbHideTitleBar.isChecked()
    self.settings.time = int(self.ui.leMinutes.text()) * \
        60 + int(self.ui.leSeconds.text())
    self.settings.save()
    self.accept()

  def close(self) -> None:
    self.reject()

File: Pomodoro/settingswindow.cpp
#include "settingswindow.h"

SettingsWindow::SettingsWindow(QWidget *parent)
	: QDialog(parent)
{
	ui.setupUi(this);

	setWindowFlag(Qt::FramelessWindowHint);
	setAttribute(Qt::WA_TranslucentBackground);
	setAttribute(Qt::WA_NoSystemBackground);

	load_settings();
}

SettingsWindow::~SettingsWindow()
{}

void SettingsWindow::on_btnSave_clicked()
{
	QString GROUP = "SettingsWindow";
	int pomo_time = ui.sbPMinutes->value() * 60 + ui.sbPSeconds->value();
	int rest_time = ui.sbRMinutes->value() * 60 + ui.sbRSeconds->value();

	Settings::save_settings("pomo_time", pomo_time, GROUP);
	Settings::save_settings("rest_time", rest_time, GROUP);
	Settings::save_settings("hide_title_bar", ui.cbAutoHideTitleBar->isChecked(), GROUP);
	Settings::save_settings("auto_start_rest_time", ui.cbAutoStartRestTime->isChecked(), GROUP);

	this->accept();
}

void SettingsWindow::on_btnCancel_clicked()
{
	this->reject();
}

void SettingsWindow::load_settings()
{
	QString GROUP = "SettingsWindow";

	int pomo_time = Settings::load_settings("pomo_time", 1500, GROUP).value<int>();
	int rest_time = Settings::load_settings("rest_time", 300, GROUP).value<int>();
	bool hide_title_bar = Settings::load_settings("hide_title_bar", false, GROUP).value<bool>();
	bool auto_start_rest_time = Settings::load_settings("auto_start_rest_time", false, GROUP).value<bool>();
	
	ui.sbPMinutes->setValue(pomo_time / 60);
	ui.sbPSeconds->setValue(pomo_time % 60);
	ui.sbRMinutes->setValue(rest_time / 60);
	ui.sbRSeconds->setValue(rest_time % 60);
	ui.cbAutoHideTitleBar->setChecked(hide_title_bar);
	ui.cbAutoStartRestTime->setChecked(auto_start_rest_time);
}
